fairing/train.py

```diff
@@ -148,26 +156,26 @@ class UserClass(cls):
             def __init__(user_class):
                 user_class.is_training_initialized = False
 
-            def __getattribute__(user_class, attribute_name):
+            def __getattribute__(user_class, attribute_name, *args, **kwargs):
                 # Overriding train in order to minimize the changes necessary in the user
                 # code to go from local to remote execution.
                 # That way, by simply commenting or uncommenting the Train decorator
                 # Model.train() will execute either on the local setup or in kubernetes
 
-                if attribute_name != 'train' or user_class.is_training_initialized:
+                if attribute_name != 'train' or user_class.is_training_initialized or not is_in_notebook():
```

Comment:
Can you explain why you made this change?
Presumably this will break launching trainings from regular python files?

Reply:
Yes. I am not sure `is_training_initialized` is working. The idea of it is sound: if training is initialized we should execute methods of this class.
if training is not initialized, execute `deploy_training`(presumably from a notebook/local env) which spawns pods in distributed mode. Then training is marked initialized, but I think that state just remains in local env, and does not carry over to the distributed env. What will tell code in distributed env (pod/container) that training has been initialized?

Reply:
> What will tell code in distributed env (pod/container) that training has been initialized?
If the code is in the remote environment where it should start training (I call that `runtime` in the code), then `is_runtime_phase()` will evaluate to true.

Consequently, `if attribute_name != 'train' or user_class.is_training_initialized` == `False`
And, `if attribute_name == 'train' and not is_runtime_phase()` == `False`.

So on the first pass (when the container starts) `is_training_initialized` will be set to `True`, the training will start, and every following call to `__getattribute__()` will resolve correctly.

Not that for `is_runtime_phase()` to evaluate to `False`, the `FAIRING_RUNTIME` environment varibable must be set, which is done here: https://github.com/wbuchwalter/fairing/blob/master/fairing/builders/dockerfile.py#L51.
Since you are not using a dockerfile, you will need to find another way to set this environment variable with config maps.

Reply:
I got it all working. I did have to return the actual function being called. However, this can still be wrapped up with the hp and pbt based wrappers.